--- libc-orig/elf/dynamic-link.h	2004-09-30 10:44:25.562284288 -0400
+++ libc/elf/dynamic-link.h	2004-09-30 10:43:05.257492472 -0400
@@ -62,6 +62,9 @@
 # define VERSYMIDX(sym)	(DT_NUM + DT_THISPROCNUM + DT_VERSIONTAGIDX (sym))
 #endif
 
+#ifndef ELF_MACHINE_SIZEOF_JMP_SLOT
+# define ELF_MACHINE_SIZEOF_JMP_SLOT sizeof(ElfW(Addr))
+#endif
 
 /* Read the dynamic section at DYN and fill in INFO with indices DT_*.  */
 #ifndef RESOLVE
