--- libc-orig/sysdeps/hppa/dl-machine.h	2004-04-18 17:17:55.000000000 -0400
+++ libc/sysdeps/hppa/dl-machine.h	2004-09-06 01:18:03.000000000 -0400
@@ -181,7 +181,7 @@
   
   extern void _dl_runtime_resolve (void);
   extern void _dl_runtime_profile (void);
-  
+ 
   /* Linking lazily */
   if (lazy)
     {
@@ -215,9 +215,10 @@
 	    {
               /* Found the GOT! */       	
               register Elf32_Addr ltp __asm__ ("%r19");
-              /* Identify this shared object. */
+              
+              /* Identify this shared object. Second entry in the got. */
               got[1] = (Elf32_Addr) l;
-
+              
               /* This function will be called to perform the relocation. */
               if (__builtin_expect (!profile, 1))
                 {
@@ -471,7 +472,7 @@
 "	ldw	4(%r3),%r19\n"	/* load the object's gp */		\
 "	bv	%r0(%r2)\n"						\
 "	depi	2,31,2,%r23\n"	/* delay slot */			\
-	);
+);
 
 
 /* This code gets called via the .plt stub, and is used in
