From c75204f1b7ae7b080c506d4b0c4179b2b40ffa62 Mon Sep 17 00:00:00 2001
From: Ulrich Drepper <drepper@redhat.com>
Date: Thu, 15 May 2008 01:56:43 +0000
Subject: [PATCH] (do_test): Use de_DE.UTF-8 locale.

---
 iconvdata/bug-iconv6.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/iconvdata/bug-iconv6.c b/iconvdata/bug-iconv6.c
index f920954..b364f4c 100644
--- a/iconvdata/bug-iconv6.c
+++ b/iconvdata/bug-iconv6.c
@@ -13,7 +13,7 @@ static const char testbuf[] = {
 static int
 do_test (void)
 {
-  setlocale (LC_ALL, "en_US.UTF-8");
+  setlocale (LC_ALL, "de_DE.UTF-8");
   iconv_t ic = iconv_open ("ISO-2022-JP//TRANSLIT", "UTF-8");
   if (ic == (iconv_t) -1)
     {
-- 
1.6.0.4

